README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# nfzRapi

<!-- badges: start -->
<!-- badges: end -->

nfzRapi is a set of functions to get data from
`https://api.nfz.gov.pl/app-umw-api/` - an API of Narodowy Fundusz
Zdrowia. As for September 2022 availabe functions let the user to get
general data about agreements and providers.

## Installation

You can install the development version of nfzRapi from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("strebuh/nfzRapi")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(nfzRapi)

# get available years
agr_get_years()
#>  [1] 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2018 2019 2020 2021 2022

## get agreements of providers in Łosice, Mazovian voivodship in 2020
df = agr_get_agreements(year=2020, admin_branch='07', town="Łosice", service_type='03')
#> [1] "https://api.nfz.gov.pl/app-umw-api/agreements?year=2020&branch=07&serviceType=03&place=Losice&limit=1&format=json&api-version=1.2"
#> No availabe data for queried scope. Please make sure the arguements are correct.
#> If any of the provided arguments contains Polish characters make sure the local encoding is one of 1250 variations. Run Sys.getlocale('LC_CTYPE') command; if it doesn't show any of 1250's You may call Sys.setlocale('LC_CTYPE', 'Polish_Poland.1250') command.


## get detail about providers in the same year and town
df2 = agr_get_providers(year=2020, admin_branch='07', town="Łosice")
#> [1] "https://api.nfz.gov.pl/app-umw-api/providers?year=2020&branch=07&place=Losice&limit=1&format=json&api-version=1.2"
#> No availabe data for queried scope. Please make sure the arguements are correct.
#> If any of the provided arguments contains Polish characters make sure the local encoding is one of 1250 variations. Run Sys.getlocale('LC_CTYPE') command; if it doesn't show any of 1250's You may call Sys.setlocale('LC_CTYPE', 'Polish_Poland.1250') command.
```


NEWS.md
# nfzRapi 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(nfzRapi)

test_check("nfzRapi")


tests/testthat/test-general_func.R

test_that("check_req_args wrong argument type or value error", {

  expect_error(check_req_args(gen="character"), "A stack number of an active frame/enviornemnt must be integer.")
  expect_error(check_req_args(gen=-1), "A stack number of an active frame/enviornemnt can't be negative.")
  expect_error(check_req_args(gen=1.5), "A stack number of an active frame/enviornemnt must be integer.")

})

# https://stackoverflow.com/questions/8771942/how-can-i-reference-the-local-environment-within-a-function-in-r

test_that("convert_pl_signs wrong argument type warning", {

  expect_warning(convert_pl_signs(char=-12), "Non-character vector given.")

})


test_that("convert_pl_signs corectness of conversion", {

  expect_identical(convert_pl_signs(char="Ą Ć Ę Ł Ń Ó Ś Ź Ż ą ć ę ł ń ó ś ź ż"),
                 paste0("%C4%84%20%C4%86%20%C4%98%20%C5%81%20%C5%83%20%C5%93%20%C5%9A%20%C5%B9%20%C5%BB",
                        "%20%C5%85%20%C5%87%20%C5%99%20%C5%82%20%C5%84%20%C3%B3%20%C5%9B%20%C5%BA%20%C5%BC"))

})

# test_that("check_env_lang language settings check", {
#
#   org_lang = Sys.getlocale("LC_CTYPE")
#   Sys.setlocale("LC_CTYPE", "Hungarian_Hungary.1250")
#   expect_error(check_env_lang(), "The character set defined by LC_CTYPE environemt variable does not support Polish non ASCII characters.")
#   Sys.setlocale("LC_CTYPE", org_lang)
#
# })



# test_that("get_request wrong context call error", {
#
#   expect_error(get_request(available_args=list(year=2020, ow=6, town='Siedlce'), api_type="app-umw-api", schema="agreements",
#                            given_args=list(call_func='agr_get_agreements', formal_args=c('year', 'ow'), given_args=c(year=2020, ow=6))),
#                "Wrong context call; get_request should be called within one of higher level functions designed for communication with an API.")
#
# })



tests/testthat/test-agreements_api_func.R
# source("R/agreements_api_func.R")
# source("R/general_func.R")


# if running test code getst to
test_that("agr_get_agreements wrong argument error", {

  # year, admin_branch, service_type, product_code, provider_code, provider_name town, regon
  expect_error(agr_get_agreements(year=2020, admin_branch='07', unexcpected_arg=3), "unused argument")

  # agr_get_agreements(year=2020, admin_branch='07', town="Huszlew")
  for(wrong in list(18, 'string', 2025)){
    expect_error(agr_get_agreements(year=wrong, admin_branch='07'), "Wrong format or type of argument\\(s\\):")
  }


  for(wrong in list(2, 127, 'string', '')){
    expect_error(agr_get_agreements(year=2020, admin_branch=wrong), "Wrong format or type of argument\\(s\\):")
  }


  for(wrong in list(3, 777, '312')){
    expect_error(agr_get_agreements(year=2020, admin_branch='07', service_type=wrong), "Wrong format or type of argument\\(s\\):")
  }


  # agr_get_agreements(year=2020, admin_branch='07', service_type="03", product_code="03.4000.030.02",  town="Siedlce")
  for(wrong in list("03.4000", "03.40000.30.02", 03.40000)){
    expect_error(agr_get_agreements(year=2020, admin_branch='07', service_type="03", product_code=wrong), "Wrong format or type of argument\\(s\\):")
  }


  # agr_get_agreements(year=2020, admin_branch='07', service_type="03", product_code="03.4000.030.02", provider_code="70603563")
  for(wrong in list(70603563, "some string")){
    expect_error(agr_get_agreements(year=2020, admin_branch='07', service_type="03", product_code="03.4000.030.02", provider_code=wrong), "Wrong format or type of argument\\(s\\):")
  }


  # agr_get_agreements(year=2020, admin_branch='07', service_type="03", product_code="03.4000.030.02", provider_name="SZPITAL", town="siedlce")
  for(wrong in list(222)){
    expect_error(agr_get_agreements(year=2020, admin_branch='07', service_type="03", product_code="03.4000.030.02", provider_code=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_agreements(year=2020, admin_branch='07', service_type="03", town="Łosice")
  # agr_get_agreements(year=2020, admin_branch='03', service_type="03", town="Biała Podlaska")
  for(wrong in list(777)){
    expect_error(agr_get_agreements(year=2020, admin_branch='07', service_type="03", town=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_agreements(year=2020, admin_branch='03', service_type="03", regon='000676708') # can be a part
  # agr_get_agreements(year=2020, admin_branch='03', service_type="03", regon='676708')
  for(wrong in list('abd', '777456456654777', 'some string')){
    expect_error(agr_get_agreements(year=2020, admin_branch='07', service_type="03", regon=wrong), "Wrong format or type of argument\\(s\\):")
  }

})


test_that("agr_get_agreements wrong argument messsage or warning", {

  expect_message(agr_get_agreements(year=2020, admin_branch='07', service_type="03", product_code="03.4000.030.02", town="siedlce", provider_name="some string"), "No availabe data for queried scope.")
  expect_message(agr_get_agreements(year=2020, admin_branch='07', service_type="03", town="Biała-Podlaska"), "No availabe data for queried scope.")
  expect_message(agr_get_agreements(year=2020, admin_branch='07', service_type="03", town="some string"), "No availabe data for queried scope.")

})



test_that("agr_get_providers wrong argument error", {

# year, admin_branch, provider_code, provider_name, nip, regon, post_code, street, town, teryt
expect_error(agr_get_providers(year=2020, admin_branch='07', unexcpected_arg=3), "unused argument")

for(wrong in list(18, 2025, 'some sting')){
  expect_error(agr_get_providers(year=wrong, admin_branch='07'), "Wrong format or type of argument\\(s\\):")
}

# agr_get_providers(year=2020, admin_branch='07')
for(wrong in list(2, 777, 'string', '')){
  expect_error(agr_get_providers(year=2020, admin_branch=wrong), "Wrong format or type of argument\\(s\\):")
}

# agr_get_providers(year=2020, admin_branch='07', provider_code="70603563")
for(wrong in list(70603563, 'some string')){
  expect_error(agr_get_providers(year=2020, admin_branch='07', provider_code=wrong), "Wrong format or type of argument\\(s\\):")
}

# agr_get_providers(year=2020, admin_branch='07', provider_name="SZPITAL", town="siedlce")
for(wrong in list(222)){
  expect_error(agr_get_providers(year=2020, admin_branch='07', town="siedlce", provider_name=wrong), "Wrong format or type of argument\\(s\\):")
}

# agr_get_providers(year=2020, admin_branch='07', nip="8212577607") # can be a part
for(wrong in list("789456", "774564566547", "some string")){
  expect_error(agr_get_providers(year=2020, admin_branch='07', town="siedlce", nip=wrong), "Wrong format or type of argument\\(s\\):")
}

# agr_get_providers(year=2020, admin_branch='03', regon=30206988) # can be a part
for(wrong in list("abd", "777456456654777", "some string")){
  expect_error(agr_get_providers(year=2020, admin_branch='07', town="siedlce", regon=wrong), "Wrong format or type of argument\\(s\\):")
}

# agr_get_providers(year=2020, admin_branch='03', post_code=2150)
for(wrong in list(215000, "21-50", "some string")){
  expect_error(agr_get_providers(year=2020, admin_branch='03', post_code=wrong), "Wrong format or type of argument\\(s\\):")
}

# agr_get_providers(year=2020, admin_branch='07', town="Huszlew")
for(wrong in list(150000)){
  expect_error(agr_get_providers(year=2020, admin_branch='07', town=wrong), "Wrong format or type of argument\\(s\\):")
}

  # agr_get_providers(year=2020, admin_branch='07', teryt=1410)
  for(wrong in list('some string', 14650111, "???")){
    expect_error(agr_get_providers(year=2020, admin_branch='07', teryt=wrong), "Wrong format or type of argument\\(s\\):")
  }
})


test_that("agr_get_providers wrong argument message or warning", {

  # agr_get_providers(year=2020, admin_branch='07', post_code=82, street="-")
  expect_message(agr_get_providers(year=2020, admin_branch='07', town="siedlce", provider_name="some string"), "No availabe data for queried scope.")
  expect_message(agr_get_providers(year=2020, admin_branch='07', town="Siedlce", street="15000000000"), "No availabe data for queried scope.")
  expect_message(agr_get_providers(year=2020, admin_branch='07', town="Siedlce", street="???"), "No availabe data for queried scope.")

})




test_that("agr_get_prov_by_year wrong argument error", {

  # year, admin_branch, provider_code, provider_name, nip, regon, post_code, street, town, teryt
  expect_error(agr_get_prov_by_year(year=2020, admin_branch='07', unexcpected_arg=3), "unused argument")

  for(wrong in list(18, 2025, 'some sting')){
    expect_error(agr_get_prov_by_year(year=wrong, admin_branch='07'), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_prov_by_year(year=2020, admin_branch='07')
  for(wrong in list(2, 123, 'string', '')){
    expect_error(agr_get_prov_by_year(year=2020, admin_branch=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_prov_by_year(year=2020, admin_branch='07', provider_code="70603563")
  for(wrong in list(70603563, 'some string')){
    expect_error(agr_get_prov_by_year(year=2020, admin_branch='07', provider_code=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_prov_by_year(year=2020, admin_branch='07', provider_name="SZPITAL", town="siedlce")
  for(wrong in list(222)){
    expect_error(agr_get_prov_by_year(year=2020, admin_branch='07', town="siedlce", provider_name=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_prov_by_year(year=2020, admin_branch='07', nip="8212577607") # can be a part
  for(wrong in list("789456", "774564566547", "some string")){
    expect_error(agr_get_prov_by_year(year=2020, admin_branch='07', town="siedlce", nip=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_prov_by_year(year=2020, admin_branch='03', regon=30206988) # can be a part
  for(wrong in list("abd", "777456456654777", "some string")){
    expect_error(agr_get_prov_by_year(year=2020, admin_branch='07', town="siedlce", regon=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_prov_by_year(year=2020, admin_branch='03', post_code=2150)
  for(wrong in list(215000, "21-50", "some string")){
    expect_error(agr_get_prov_by_year(year=2020, admin_branch='03', post_code=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_prov_by_year(year=2020, admin_branch='07', town="Huszlew")
  for(wrong in list(150000)){
    expect_error(agr_get_prov_by_year(year=2020, admin_branch='07', town=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_prov_by_year(year=2020, admin_branch='07', teryt=1410)
  for(wrong in list('some string', 14650111, "???")){
    expect_error(agr_get_prov_by_year(year=2020, admin_branch='07', teryt=wrong), "Wrong format or type of argument\\(s\\):")
  }

  })

test_that("agr_get_prov_by_year wrong argument mesage or warning", {

  # agr_get_prov_by_year(year=2020, admin_branch='07', post_code=82, street="-")
  expect_message(agr_get_prov_by_year(year=2020, admin_branch='07', town="siedlce", provider_name="some string"), "No availabe data for queried scope.")
  expect_message(agr_get_prov_by_year(year=2020, admin_branch='07', town="Siedlce", street="15000000000"), "No availabe data for queried scope.")
  expect_message(agr_get_prov_by_year(year=2020, admin_branch='07', town="Siedlce", street="???"), "No availabe data for queried scope.")
  expect_message(agr_get_prov_by_year(year=2020, admin_branch='07', street="???"), "No availabe data for queried scope.")

})



test_that("agr_get_provider wrong argument error", {

  # agr_get_provider(year, provider_code, admin_branch)
  expect_error(agr_get_provider(year=2020, admin_branch='07', unexcpected_arg=3), "unused argument")

  for(wrong in list(18, 2025, 'some sting')){
    expect_error(agr_get_provider(year=wrong, admin_branch='07', provider_code='70603563'), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_provider(year=2020, admin_branch='07', provider_code='70603563')
  for(wrong in list(2, 777, 'string', '')){
    expect_error(agr_get_provider(year=2020, admin_branch=wrong, provider_code='70603563'), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_provider(year=2020, admin_branch='07', provider_code="70603563")
  for(wrong in list(70603563, 'some string')){
    expect_error(agr_get_provider(year=2020, admin_branch='07', provider_code=wrong), "Wrong format or type of argument\\(s\\):")
  }

})


test_that("agr_get_serivces wrong argument error", {

  # agr_get_serivces(year, service_code=NULL, service_name=NULL)
  expect_error(agr_get_serivces(year=2020, service_code='04', unexcpected_arg=3), "unused argument")

  for(wrong in list(18, 2025, 'some sting')){
    expect_error(agr_get_serivces(year=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_serivces(year=2020, service_code='07')
  for(wrong in list(123, 'string', '')){
    expect_error(agr_get_serivces(year=2020, service_code=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_serivces(year=2020, admin_branch='07', provider_name="SZPITAL", town="siedlce")
  for(wrong in list(222)){
    expect_error(agr_get_serivces(year=2020, service_name=wrong), "Wrong format or type of argument\\(s\\):")
  }

})


test_that("agr_get_serivces wrong argument message or warning", {

  expect_message(agr_get_serivces(year=2020, service_name='some string'), "No availabe data for queried scope.")

})


test_that("agr_get_serivces_year wrong argument error", {

  # agr_get_serivces_year(year)
  expect_error(agr_get_serivces_year(year=2020, unexcpected_arg=3), "unused argument")

  for(wrong in list(18, 2025, 'some sting')){
    expect_error(agr_get_serivces_year(year=wrong), "Wrong format or type of argument\\(s\\):")
  }
})


test_that("agr_get_products wrong argument error", {

  # agr_get_products(year, prod_code=NULL, product_name=NULL)
  expect_error(agr_get_products(year=2020, unexcpected_arg=3), "unused argument")

  for(wrong in list(18, 2025, 'some sting')){
    expect_error(agr_get_products(year=wrong), "Wrong format or type of argument\\(s\\):")
  }

  # agr_get_products(year=2020, service_type='07')
  for(wrong in list(03.40000, "03.40000.030.02", "03.4000.3.02", "3.4000.030.02.")){
    expect_error(agr_get_products(year=2020, prod_code=wrong), "Wrong format or type of argument\\(s\\):")
  }

  for(wrong in list(03.40000)){
    expect_error(agr_get_products(year=2020, product_name=wrong), "Wrong format or type of argument\\(s\\):")
  }
})

test_that("agr_get_products wrong argument message or warning", {

  expect_warning(agr_get_products(year=2020, prod_code="03.4000.030"), "Given prod_code length is \\d+ which is")
  expect_warning(agr_get_products(year=2020, prod_code="03.4000.30.02"), "Given prod_code length is \\d+ which is")
  # TODO: check for correct format, also when part of the code provided

})








R/general_func.R


#' @title Get data from selected api type and schema provided available arguments
#' @description General function to get data from selected NFZ API type (agreements, ...) and schema (ex. agreements or providers), provided list of available arguments. Function to be run inside outer function for selected API User given arguments are retrieved from appropriate environment.
#' @param available_args Dictionary of function arguments and their corresponding equivalents in NFZ API.
#' @param api_type A type of NFZ api to be queried.
#' @param schema A name of schema of appropriate api.
#' @param url_args A vector of obligatory arguments to be passed without keys as a part of the url <https://api.nfz.gov.pl/api_type/schema/url_args/url_args>.
#' @return  Returns a data.table for requested query.
get_request <- function(available_args, api_type="app-umw-api", schema="agreements", url_args=NULL){

  schemas_dict <- list(`app-umw-api`=c('agreements', "months", "plans", "providers", "service-types", "contract-products"))
  if(!api_type %in% names(schemas_dict)){
    stop(paste0("Wrong API type, ", api_type, " doesn't exist."))
  } else if(!schema %in% schemas_dict[[api_type]]){
    stop(paste0(schema, " schema doesn't exist within ", api_type, " API."))
  }

  # Retrieve arguments and check their corectness
  given_args <- check_req_args(sys.parent(1))
  functions <- c("agr_get_agreements", "agr_get_agreement", "agr_get_plan", "agr_get_month_plan", "agr_get_providers", "agr_get_prov_by_year", "agr_get_provider",
                 "agr_get_serivces", "agr_get_products")
  if(!as.character(given_args$call_func) %in% functions){
    stop(paste("Wrong context call; get_request should be called within one of higher level functions designed for communication with an API."))
  }

  formal_names <- given_args$formal_args
  given_args <- given_args$given_args

  # Prepare query provided arguments
  formal_args = c(api_type, schema)
  if(!is.null(url_args)){
    n_formals = if(length(url_args) == length(given_args)) 0 else length(intersect(names(given_args), url_args))
    optional_args = if(length(url_args) == length(given_args)) list() else given_args[(n_formals+1):length(given_args)]
    formal_args = c(formal_args, unlist(setdiff(given_args, optional_args)))
  } else {
    optional_args = given_args
  }
  optionals_matcher <- if(rlang::is_empty(optional_args)) "" else "="
  api_agrs <- paste0(sapply(names(optional_args), function(x) eval(available_args[[x]])),
                     optionals_matcher,
                     sapply(names(optional_args), function(x) eval(optional_args[[x]])))
  api_agrs <- convert_pl_signs(api_agrs)

  api_query <- paste0("https://api.nfz.gov.pl/", paste(formal_args, collapse = "/"), "?",
                      paste0(api_agrs, collapse = "&"), "&limit=1&format=json&api-version=1.2")
  print(api_query)
  # Get scope request (to know how much data is available)
  request <- httr::GET(api_query, httr::timeout(20))
  request_data <- jsonlite::fromJSON(httr::content(request, "text"), flatten=TRUE)

  status_code <- request$status_code
  if(status_code != 200){
    msg=paste0("No availabe data for queried scope. Please make sure the arguements are correct.\n",
               "Request error, status code:", as.character(status_code))
    message(msg)
    return(NULL)
  }

  # Extract data from the request
  data <- data.table::data.table()
  if(!is.null(request_data$data)){

    items_found <- request_data$meta$count
    message(paste0(items_found, ' items of data meeting requested criteria found.'))
    api_query <- gsub("&limit=1", paste0("&page=", 1,"&limit=25"), api_query)
    # Get requests for all available data (max items per page 25 from API doc.)
    for(page in 1:ceiling(items_found/25)){

      api_query <- gsub("&page=\\d+&limit=25", paste0("&page=", page,"&limit=25"), api_query)
      request <- httr::GET(api_query, httr::timeout(20))
      request_data <- jsonlite::fromJSON(httr::content(request, "text"), flatten=TRUE)
      data_attribute = sapply(request_data$data, class)
      data_attribute = match("data.frame", data_attribute)
      data <- rbind(data, request_data$data[[data_attribute]])
      page <- page + 1
      Sys.sleep(0.1)

    }
  } else {
    msg=paste0("No availabe data for queried scope. Please make sure the arguements are correct.\n",
                 "If any of the provided arguments contains Polish characters make sure the local encoding is one of 1250 variations. Run Sys.getlocale('LC_CTYPE') command; if it doesn't show any of 1250's ",
                 "You may call Sys.setlocale('LC_CTYPE', 'Polish_Poland.1250') command.")
    message(msg)
    # browser()
    return(NULL)
  }
  # browser()
  return(data)
}



#' @title Check for function's arguments
#' @description To be run in local environment (inside another function). Checks whether all required arguments of function run in global env were provided during a function call. If a vector of possible arguments was given also checks whether all provided in a function call arguments are supported.
#' @param gen The stack number of a frame (environment) which arguments are to be verified
#' @return  Returns a vector of arguments provided in a function call if all arguments are correct, otherwise function stops
#' @details `gen` describes a relative position or a stack number of check_req_args parent environment relative to a global environment (global has number 0). Can't be 0 or negative.
#' @examples
#' \dontrun{
#'  foo <- function(a, b, ...){
#'    given_args = check_req_args(foo, c("a", "b", "c"))
#'    return(given_args)}
#'
#'  foo("a", "b", "c", "d")
#' }
#'
check_req_args <- function(gen){


  if(!is.numeric(gen)){
    stop("A stack number of an active frame/enviornemnt must be integer.")
  } else if(gen!=round(gen)){
    stop("A stack number of an active frame/enviornemnt must be integer.")
  }
  if(gen < 0){
    stop("A stack number of an active frame/enviornemnt can't be negative.")
  }

  func_call <- as.list(match.call(definition = sys.function(gen), # stack number of api function env (usually 1, but if nested into antohter funtion not 1)
                                  call = sys.call(gen), # sys.call(sys.parent(gen)) # stack number of api function env (usually 1, but if nested into antohter funtion not 1)
                                  envir = sys.frame(gen))) # parent.frame(gen) -gen

  given_args <- func_call[2:length(func_call)]
  # !!!This won't pass test when test's run in a loop in devtools, because iterated element't is hidden in some env  and can't be evaluated
  gen.2 = gen
  while("name" %in% sapply(given_args, class)){
    gen.2 <-  gen.2-1
    given_args <- lapply(given_args, function(e) ifelse(is.name(e), eval(e, envir=sys.frame(gen.2)), e))
  }
  # Get arguments that are required, and don't have default value
  possible_args <- formals(fun = sys.function(gen), envir = sys.frame(gen))
  required_args <- names(unlist(possible_args))# While unlisting a pairList arguments with default NULL are dropped out
  possible_args <- names(possible_args)

  missing_args <- setdiff(required_args, names(given_args))
  if(!rlang::is_empty(missing_args)){
    msg <- paste0("You did not provide all reqired arguments. Missing: ", paste(missing_args, collapse = ", "))
    stop(msg)
  }

  return(list(call_func = func_call[[1]], formal_args=required_args, available_args=possible_args, given_args=given_args))
}


agr_check_arg_types <- function(year=NULL, admin_branch=NULL, service_type=NULL, service_code=NULL, service_name=NULL, product_code=NULL, prod_code=NULL,
                                provider_code=NULL, provider_name=NULL, product_name=NULL, nip=NULL, regon=NULL, post_code=NULL, street=NULL, town=NULL,
                                teryt=NULL, id_agreement=NULL, id_plan=NULL){


  err = "Wrong format or type of argument(s):"
  msg = ""
  if(!is.null(year)){
    if(!nchar(year) == 4){
      err = paste(err, "year - 4 digits expected.", sep="\n")
    } else if(!year %in% agr_get_years()){
      err = paste(err, "Year out of range. To check available years call agr_get_years function.", sep="\n")
    }
  }

  if(!is.null(admin_branch)){
    if(grepl("\\D", admin_branch) | admin_branch=="" | !is.character(admin_branch)){
      err = paste(err, "admin_branch - A code should be a number between 1 and 16 (for precise results 2 digits character including leading zero).", sep="\n")
    } else if(nchar(admin_branch) > 2){
      err = paste(err, "admin_branch - out of range. A code should be a number between 1 and 16 (for precise results 2 digits character including leading zero).", sep="\n")
    }
  }

  if(!is.null(service_type)){
    if(!is.null(year)){
      possible_services = agr_get_serivces_year(year)$attributes.code
      if(!nchar(service_type) == 2){
        err = paste(err, "service_type - 2 digits character code expected (including leading zero if needed).", sep="\n")
      } else if(!service_type %in% possible_services){
        err = paste(err, paste0("service_type - out of range. Possible values are: ", paste(possible_services, collapse = ", ")), sep="\n")
      }
    } else {
      stop("year missing.")
    }
  }

  if(!is.null(service_code)){
    if(grepl("\\D", service_code) | service_code==""){
      err = paste(err, "service_code - 2 digits character code expected (including leading zero if needed).", sep="\n")
    } else if(nchar(service_code) > 2){
      err = paste(err, "service_code - 2 digits character code expected (including leading zero if needed).", sep="\n")
    } else if(nchar(service_code) < 2){
      msg = paste(msg, paste0("Given service_code length is ", nchar(service_code)," which may be ambiguous. Results returned for all codes that match given subset."), sep="\n")
    }
  }

  if(!is.null(service_name)){
    if(!is.character(service_name)){
      err = paste(err, "service_name - character expected.", sep="\n")
    }
  }

  if(!is.null(product_code)){
    if(!is.character(product_code) | !grepl("\\d{2}\\.\\d{4}\\.\\d{3}\\.\\d{2}", product_code)){
      err = paste(err, "product_code - expected full code of 11 digits separated by dots, format: 'dd.dddd.ddd.dd'.", sep="\n")
    }
  }

  if(!is.null(prod_code)){
    if(!is.character(prod_code) | grepl("(\\.(\\d{1}|\\d{5,})\\.)|((.*\\..*){4,})", prod_code)){
      err = paste(err, "prod_code - character code of 11 digits separated by dots, full code correct format is: 'dd.dddd.ddd.dd' (subset can be provided).", sep="\n")
    } else if(nchar(prod_code)>14){
      err = paste(err, paste0("Given prod_code length is ", nchar(prod_code)," which is too long, full code correct format is: 'dd.dddd.ddd.dd' (subset can be provided)."), sep="\n")
    } else if(nchar(prod_code)<14){
      msg = paste(msg, paste0("Given prod_code length is ", nchar(prod_code)," which is not a full code, full code correct format is: 'dd.dddd.ddd.dd' (subset can be provided)."), sep="\n")
    }
  }

  if(!is.null(product_name)){
    if(!is.character(product_name)){
      err = paste(err, "product_name - character expected.", sep="\n")
    }
  }

  if(!is.null(provider_code)){
    if(!is.character(provider_code) | !grepl("\\d+", provider_code)){
      err = paste(err, "provider_code - character expected.", sep="\n")
    }
  }

  if(!is.null(provider_name)){
    if(!is.character(provider_name)){
      err = paste(err, "provider_name - character expected.", sep="\n")
    }
  }

  if(!is.null(nip)){
    if(nchar(nip)!=10 | grepl("\\D", nip)){
      err = paste(err, "nip - 10 digits code expected.", sep="\n")
    }
  }

  if(!is.null(regon)){
    if(nchar(regon)>14 | grepl("\\D", regon)){
      err = paste(err, "regon - code too long or contaning non-numeric signs. The max length of the code is 14 digits. If code's subset provided, all records that match the subset will be returned.", sep="\n")
    }
  }

  if(!is.null(post_code)){
    if(nchar(post_code)>5 | grepl("\\D", post_code)){
      err = paste(err, "post_code - 5 digits post code without a hyphen ('-') expected.", sep="\n")
    }
  }

  if(!is.null(town)){
    if(!is.character(town)){
      err = paste(err, "town - character expected.", sep="\n")
    }
  }

  if(!is.null(teryt)){
    if(nchar(teryt)>7 | grepl("\\D", teryt)){
      err = paste(err, "teryt - should consist of max 7 digits.", sep="\n")
    }
  }

  if(!is.null(id_agreement)){
    if(!is.character(id_agreement)){
      err = paste(err, "id_agreement - character expected.", sep="\n")
    }
  }

  if(!is.null(id_plan)){
    if(!is.character(id_plan)){
      err = paste(err, "id_plan - character expected.", sep="\n")
    }
  }

  if(msg!=""){
    warning(msg)
  }

  if(err!="Wrong format or type of argument(s):"){
    stop(err)
  }

}


#' @title Check for compatibility of LC_CTYPE environment variable
#' @description To be able to download data from NFZ api service LC_CTYPE env. var. has to be set to Polish language. This function checks for that, and suggest code to run, to be able to use NFZ api. If LC_CTYPE doesn't match, function stops the call.
check_env_lang <- function(){
  char_set <- Sys.getlocale("LC_CTYPE")
  if(char_set != "Polish_Poland.1250"){
    msg = paste0("Your LC_CTYPE environment variable has the value '", char_set,"', but to be sure that nfzRapi functions work correctly it has to be changed to 'Polish_Poland.1250'.\n",
    "1. Before using nfzRapi functions run following command: Sys.setlocale('LC_CTYPE', 'Polish_Poland.1250')\n",
    "2. After using nfzRapi, run following command to restore original setting: Sys.setlocale('LC_CTYPE', '", char_set,"')")
    message(msg)
    stop("The character set defined by LC_CTYPE environemt variable does not support Polish non ASCII characters.")
  }
}



#' @title Convert Polish non ASCII characters to NFZ API accpted utf-8 representation of Unicode
#' @description Function accpets a string and converts Polish diacretic non ASCII characters
#' @param char A character where the signs are to be converted
#' @return  Returns a character vector
#' @examples
#' \dontrun{
#'  convert_pl_signs("Gąśłękówiżań")
#' }
#'
convert_pl_signs <- function(char){


  if(!is.character(char)){
    warning("Non-character vector given.")
  }

  unicode_dict <- list(
    `%C4%84`="\u104",
    `%C4%86`="\u106",
    `%C4%98`="\u118",
    `%C5%81`="\u141",
    `%C5%83`="\u143",
    `%C5%93`="\uD3",
    `%C5%9A`="\u15A",
    `%C5%B9`="\u179",
    `%C5%BB`="\u17B",
    `%C5%85`="\u105",
    `%C5%87`="\u107",
    `%C5%99`="\u119",
    `%C5%82`="\u142",
    `%C5%84`="\u144",
    `%C3%B3`="\uF3",
    `%C5%9B`="\u15B",
    `%C5%BA`="\u17A",
    `%C5%BC`="\u17C",
    `%20`=" "
  )
  char2 <- c()
  for(character in char){
    for(letter in unicode_dict){
      change <- names(unicode_dict)[match(letter, unicode_dict)]
      character <- gsub(letter, change[!is.na(change)], character)
    }
    char2 <- c(char2, character)
  }
  return(char2)
}


